File: src/algorithms/trees/properties/binary-tree-tilt/sources/binary-tree-tilt_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("binary-tree-tilt")
TreeNode = module.TreeNode
binary_tree_tilt = module.binary_tree_tilt


def make_node(value, left=None, right=None):
    node = TreeNode(value)
    node.left = left
    node.right = right
    return node


def test_null_root_returns_zero():
    assert binary_tree_tilt(None) == 0


def test_single_node_returns_zero():
    assert binary_tree_tilt(make_node(1)) == 0


def test_simple_3_node_tree():
    # tilt at root = |2 - 3| = 1, leaves have tilt 0, total = 1
    assert binary_tree_tilt(make_node(1, make_node(2), make_node(3))) == 1


def test_non_negative_for_any_tree():
    root = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    assert binary_tree_tilt(root) >= 0


if __name__ == "__main__":
    test_null_root_returns_zero()
    test_single_node_returns_zero()
    test_simple_3_node_tree()
    test_non_negative_for_any_tree()
    print("All tests passed!")

File: src/algorithms/trees/properties/binary-tree-tilt/binary-tree-tilt_test.rs
include!("sources/binary-tree-tilt.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<TreeNode>>, right: Option<Box<TreeNode>>) -> Option<Box<TreeNode>> {
        Some(Box::new(TreeNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<TreeNode>> {
        make_node(value, None, None)
    }

    #[test]
    fn test_null_root_returns_zero() {
        assert_eq!(binary_tree_tilt(&None), 0);
    }

    #[test]
    fn test_single_node_returns_zero() {
        assert_eq!(binary_tree_tilt(&leaf(1)), 0);
    }

    #[test]
    fn test_simple_3_node_tree() {
        let root = make_node(1, leaf(2), leaf(3));
        assert_eq!(binary_tree_tilt(&root), 1);
    }

    #[test]
    fn test_non_negative_for_any_tree() {
        let root = make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)));
        assert!(binary_tree_tilt(&root) >= 0);
    }
}
